larpandoracontent/LArMonitoring/EventClusterValidationAlgorithm.cc

```diff
@@ -129,20 +198,49 @@ StatusCode EventClusterValidationAlgorithm::Run()
 //------------------------------------------------------------------------------------------------------------------------------------------
 
 void EventClusterValidationAlgorithm::GetHitParents(
-    const CaloHitList &caloHits, const ClusterList &clusters, std::map<const CaloHit *const, CaloHitParents> &hitParents) const
+        const CaloHitList &caloHits, const ClusterList &clusters, std::map<const CaloHit *const, CaloHitParents> &hitParents) const
 {
+    std::map<const MCParticle *const, const MCParticle *const> mcFoldTo;
     for (const CaloHit *const pCaloHit : caloHits)
     {
+        MCParticleWeightMap weightMap{pCaloHit->GetMCParticleWeightMap()};
+
+        if(m_foldShowers)
+        {
+            MCParticleWeightMap foldedWeightMap;
+            for (const auto &[pMC, weight] : weightMap)
+            {
+                const MCParticle *pFoldedMC{nullptr};
+                if (mcFoldTo.find(pMC) != mcFoldTo.end())
+                {
+                    pFoldedMC = mcFoldTo.at(pMC);
+                }
+                else
+                {
+                    pFoldedMC = this->FoldMCTo(pMC);
+                    mcFoldTo.insert({pMC, pFoldedMC});
+                }
+                foldedWeightMap[pFoldedMC] += weight;
+            }
+            weightMap = foldedWeightMap;
+        }
```

Comment:
Could you compute an MC to MC map at the start of the alg and just use this to look up the approriate MC? i.e. an unfolded map would just have each MC point to itself, while the folded shower map would have MC point to itself for non-shower MC, and then point to the leading MC for the shower MC? That way you separate the folding logic from the hit to MC logic.

Reply:
Yes I see what you mean, I considered a similar implementation but I think I prefer the way it is currently.